File: app/src/main/java/com/rayworks/droidweekly/viewmodel/ViewModelFactory.java
package com.rayworks.droidweekly.viewmodel;

import android.app.Application;

import androidx.annotation.NonNull;
import androidx.lifecycle.ViewModel;
import androidx.lifecycle.ViewModelProvider;

import com.rayworks.droidweekly.repository.ArticleRepository;

public class ViewModelFactory extends ViewModelProvider.AndroidViewModelFactory {
    private final ArticleRepository repository;

    public ViewModelFactory(@NonNull Application application, ArticleRepository repository) {
        super(application);
        this.repository = repository;
    }

    @NonNull
    @Override
    public <T extends ViewModel> T create(@NonNull Class<T> modelClass) {
        if (modelClass.isAssignableFrom(ArticleListViewModel.class)) {
            return (T) new ArticleListViewModel(repository);
        }

        throw new UnsupportedOperationException(
                "No view model matched for " + modelClass.getSimpleName());
    }
}

File: app/src/main/java/com/rayworks/droidweekly/viewmodel/ViewModelFactory.kt
package com.rayworks.droidweekly.viewmodel

import android.os.Bundle
import androidx.lifecycle.AbstractSavedStateViewModelFactory
import androidx.lifecycle.SavedStateHandle
import androidx.lifecycle.ViewModel
import androidx.savedstate.SavedStateRegistryOwner
import com.rayworks.droidweekly.repository.ArticleRepository

class ViewModelFactory(owner: SavedStateRegistryOwner,
                       defaultState: Bundle?,
                       private val repository: ArticleRepository)
    : AbstractSavedStateViewModelFactory(owner, defaultState) {

    @Suppress("UNCHECKED_CAST")
    override fun <T : ViewModel?> create(key: String, modelClass: Class<T>, handle: SavedStateHandle): T {
        return ArticleListViewModel(handle, repository) as T
    }

}